index.js

```diff
@@ -0,0 +1,282 @@
+                // already part of globalModulesSet, currently our filters
+                // produce a disjoint set of modules... or at the very least
+                // warn when there an overlap in the set of entry modules
+                // between shared chunks.
+                const moduleFilter = spec.moduleFilter || (module => true);
+                if (moduleFilter(module)) {
+                    const count = commonModulesToCountMap.has(module)
+                        ? commonModulesToCountMap.get(module)
+                        : 0;
+                    commonModulesToCountMap.set(module, count + 1);
+                }
+            }
+        }
+
+        // Some shared chunks may specify a minimum number of chunks that a
+        // module must appear in before we move it to the shared chunk.
+        const minChunks = spec.minChunks || 1;
+        const commonModules = new Set();
+        for (const [module, count] of commonModulesToCountMap) {
+            if (count >= minChunks) {
+                commonModules.add(module);
+                globalModulesSet.add(module);
+            }
+        }
+
+        // Get chunk and module dependencies
+        const { moduleDeps, chunkDeps } = this.getDependencies(
+            commonModules,
+            globalModulesSet,
+            sharedChunksSet
+        );
+
+        // If there aren't any modules to add to the chunk don't bother
+        // creating a shared chunk.
+        if (commonModules.size === 0 && moduleDeps.size === 0) {
+            console.warn(
+                `'${spec.name}' chunk not created, contains no modules`
+            );
+            return;
+        }
+
+        if (allChunksNameMap.has(spec.name)) {
+            // Note: this is different from how CommonsChunkPlugin works.
+            throw new Error(
+                `SharedChunksPlugin doesn't work with existing chunks`
+            );
+        }
+
+        // This is the new shared chunk to which we'll be adding modules.
+        const sharedChunk = compilation.addChunk(spec.name);
+
+        // Add module dependencies.
+        for (const dep of moduleDeps) {
+            commonModules.add(dep);
+            globalModulesSet.add(dep);
+        }
+
+        // Add all common modules to the target chunk.
+        for (const module of commonModules) {
+            sharedChunk.addModule(module);
+            module.addChunk(sharedChunk); // modules can appear in multiple chunks
+        }
+
+        // Add chunk dependencies.
+        sharedChunk.parents = [...chunkDeps];
+        for (const chunk of chunkDeps) {
+            chunk.addChunk(sharedChunk);
+        }
+
+        // It's possible that moduleFilter function filters out all modules
+        // from a particular chunk in which case it isn't actually affected.
+        // TODO(kevinb): figure out what to do when extracting common
+        // chunks from async chunks, see CommonsChunkPlugin
```

Comment:
I left this as a TODO b/c I was waiting for the codemod diff to update `require.async` to `import`.  I'm going to do some testing once that lands and will open another PR if there are any issues that need to be addressed.